allennlp/training/trainer.py

```diff
@@ -51,11 +51,18 @@ class Trainer(Registrable):
     def __init__(
         self,
         serialization_dir: str,
-        cuda_device: Union[int, torch.device] = -1,
+        cuda_device: Union[None, int, torch.device] = None,
```

Comment:
I think `None` here might mess with the `FromParams` logic. Can you make this the last type argument, at least?

Or maybe `None` gets stripped already because of how `Optional` is handled...?  Nope, the special case there wouldn't handle this: https://github.com/allenai/allennlp/blob/aa2943e50d8ee81ea86a907c72f5d56a83ddc2fc/allennlp/common/from_params.py#L108-L109

Can you either just remove the `None` here in the type annotation, or have a test that explicitly checks that this gets the right value with various configuration options?  I'm nervous that this won't actually work as it is.

Reply:
So if I change it to `Optional[Union[int, torch.device]]`, `FromParams` will work?

Reply:
It might still work now, I just am not sure, and that makes me nervous.